code/dictionary.lisp

```diff
@@ -0,0 +1,74 @@
+(cl:in-package #:spell)
+
+;;; Dictionary
+
+(defclass dictionary (utilities.print-items:print-items-mixin)
+  ((%contents    :accessor contents
+                 :initform (make-instance 'raw-node))
+   (%entry-count :accessor entry-count
+                 :initform 0)))
+
+(defmethod make-load-form ((object dictionary) &optional environment)
+  (make-load-form-saving-slots object :environment environment))
+
+(defmethod utilities.print-items:print-items append ((object dictionary))
+  `((:entry-count "~:D entr~:@P" ,(entry-count object))))
+
+(defmethod lookup ((string string) (dictionary dictionary))
+  (let ((length (length string))
+        (result '()))
+    (check-type length (integer 1) "Query string must not be empty")
```

Comment:
`check-type` establishes a `store-value` restart which can modify `length` but not `string`, so someone silly can break the code further. It's somewhat ugly to handle that and minimize the number of `length` calls this way, but:

```lisp
(defmethod lookup ((string string) (dictionary dictionary))
  (let ((length 0)
        (result '()))
    (assert (typep (setf length (length string)) '(integer 1))
            (x) "Query string must not be empty")
    (%lookup (lambda (word)
               (push word result))
             string length (contents dictionary))
    result))
```

Reply:
I used `check-type` mainly out of laziness. I didn't really think about recovering. Would a non-recoverable `simple-error` work for you?

Reply:
Yep, just an `assert` without any slots will do in that case:
```lisp
    (assert (typep (setf length (length string)) '(integer 1))
            () "Query string must not be empty")
```

Or anything simpler like `(when (zerop length) (error ...))`.

Reply:
Done.